library/std/src/primitive_docs.rs

```diff
@@ -194,14 +194,47 @@ mod prim_bool {}
 /// # `!` and traits
 ///
 /// When writing your own traits, `!` should have an `impl` whenever there is an obvious `impl`
-/// which doesn't `panic!`. As it turns out, most traits can have an `impl` for `!`. Take [`Debug`]
+/// which doesn't `panic!`. The reason is that functions returning an `impl Trait` where `!`
+/// does not have an `impl` of `Trait` cannot diverge as their only possible code path. In other
+/// words, they can't return `!` from every code path. As an example, this code doesn't compile:
+///
+/// ```compile_fail
+/// use core::ops::Add;
+///
+/// fn foo() -> impl Add<u32> {
+///     unimplemented!()
+/// }
+/// ```
+///
+/// But this code does:
+///
+/// ```
+/// use core::ops::Add;
+///
+/// fn foo() -> impl Add<u32> {
+///     if true {
+///         unimplemented!()
+///     } else {
+///         0
+///     }
+/// }
+/// ```
+///
+/// The reason is that, in the first example, there are many possible types that `!` could coerce
+/// to, because the function can return one of many concrete types. However, in the second example,
+/// the `else` branch returns a `0` of type `u32`, which is a concrete type that `!` can be coerced
```

Comment:
I feel like this might still be a little confusing. The `else` branch doesn't explicitly say `0u32`, it's just the compiler infers `u32` from the return type. Is there a better way to say this than "a `0` of type `u32`"?

Reply:
The way you said it now seems fine: "the compiler infers that the 0 has type u32".